Cleanup StyleProperties::fontSynthesisValue()

nt1m opened the pull request:
#### 55fafd865860fec1a636313be31d403366cf9715
<pre>
Cleanup StyleProperties::fontSynthesisValue()
<a href="https://bugs.webkit.org/show_bug.cgi?id=248127">https://bugs.webkit.org/show_bug.cgi?id=248127</a>
rdar://102544812

Reviewed by Darin Adler.

- Use `propertyAsValueID` instead of custom lambda, it is possible since longhands are never added implicitly.
- Always do checks against the value ID (it is now possible now that we include the `isCSSWideKeyword()` function)
- Use pattern consistent with other methods for appending the strings to the string builder.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::fontSynthesisValue const):

Canonical link: <a href="https://commits.webkit.org/256877@main">https://commits.webkit.org/256877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5080eb936257d01dac15754db3e46e122fc4dbdc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97086 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6353 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30207 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106603 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166874 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6590 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35085 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89475 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103291 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102755 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83703 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/31957 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74844 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/376 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/358 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5157 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2321 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1599 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/40876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -717,51 +717,27 @@ String StyleProperties::fontVariantValue() const
+    if (style == CSSValueAuto)
+        result.append(result.isEmpty() ? "" : " ", "style"_s);
+    if (caps == CSSValueAuto)
+        result.append(result.isEmpty() ? "" : " ", "small-caps"_s);
```
The algorithm here is simple enough that we can do it without `StringBuilder`, just compute three strings including separators and concatenate them with `makeString`, which is more efficient.

nt1m replied:
I came up with:
```
    String weightString = weight == CSSValueAuto ? "weight"_s : ""_s;
    String styleString, capsString;
    if (style == CSSValueAuto)
        styleString = weight != CSSValueAuto ? "style"_s : " style"_s;

    if (caps == CSSValueAuto)
        capsString = weight != CSSValueAuto && style != CSSValueAuto ? "small-caps"_s : " small-caps"_s;
    return makeString(weightString, styleString, capsString);
```
but I'm not a huge fan personally.

darinadler replied:
Those can be `const char*` or `StringLiteral` or written as `auto`, no need to allocate and destroy temporary strings.

You might be able to make it read more elegant be rearranging or making a helper function. It's much more efficient to do fewer heap allocations.

OK with me either way. Your original is OK.

nt1m committed:
Cleanup StyleProperties::fontSynthesisValue()
https://bugs.webkit.org/show_bug.cgi?id=248127
rdar://102544812

Reviewed by Darin Adler.

- Use `propertyAsValueID` instead of custom lambda, it is possible since longhands are never added implicitly.
- Always do checks against the value ID (it is now possible now that we include the `isCSSWideKeyword()` function)
- Use pattern consistent with other methods for appending the strings to the string builder.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::fontSynthesisValue const):

Canonical link: https://commits.webkit.org/256877@main
